src/katgpucbf/vgpu/engine.py

```diff
@@ -62,6 +122,68 @@ def _populate_sensors(self, sensors: aiokatcp.SensorSet, send_pols: tuple[str, s
                 initial_status=aiokatcp.Sensor.Status.NOMINAL,
             )
         )
+        prefixes = [f"{pol}." for pol in self.recv_pol_labels]
+        for sensor in base_recv.make_sensors(recv_sensor_timeout, prefixes).values():
+            sensors.add(sensor)
+
+    def _init_recv(self, recv_affinity: int, monitor: Monitor) -> None:
+        """Initialise the receive side of the engine."""
+        recv_chunks = 4  # TODO: may need tuning?
+        data_ringbuffer = ChunkRingbuffer(
+            recv_chunks, name="recv_data_ringbuffer", task_name="run_receive", monitor=monitor
+        )
+        free_ringbuffer = spead2.recv.ChunkRingbuffer(recv_chunks)
+        layout = self.recv_layout
+        dtype = np.dtype(f"int{layout.sample_bits}")
+        self._recv_group = recv.make_stream_group(
+            layout, data_ringbuffer, free_ringbuffer, recv_affinity, self.recv_pol_labels
+        )
+        for _ in range(recv_chunks):
+            chunk = recv.Chunk(
+                present=np.empty(
+                    (N_POLS, layout.n_batches_per_chunk, layout.n_pol_substreams),
+                    np.uint8,
+                ),
+                data=cupyx.empty_pinned(
```

Comment:
I know this is a bit late to be asking (considering the PR to introduce cupy-cuda12x was before this), but
- What's the reasoning/requirement for pinned memory here?

Like the [documentation says](https://docs.cupy.dev/en/stable/reference/generated/cupyx.empty_pinned.html#cupyx-empty-pinned)
> This is a convenience function which is just [numpy.empty()](https://numpy.org/doc/stable/reference/generated/numpy.empty.html#numpy.empty), except that the underlying memory is pinned/pagelocked.

Reply:
Noted and thank you for that reminder - I was wondering why the layout looks similar, but the declaration is different. Follow-up question,
- Why use cupyx over katsdpsigproc.accel?

Reply:
Because that's what katcbf-vlbi-resample uses. That was a matter of expediency: it's much quicker to write something that works, and in particular it has some built-in signal processing like upfirdn that I would have had to hand-code if I was using katsdpsigproc.